src/varnish.c

```diff
@@ -127,36 +134,106 @@ static int varnish_submit_gauge(const char *plugin_instance, /* {{{ */
-  return varnish_submit(plugin_instance, category, type, type_instance,
+  return varnish_submit(plugin_instance, category, NULL, type, type_instance,
                         (value_t){
                             .gauge = (gauge_t)gauge_value,
                         });
 } /* }}} int varnish_submit_gauge */
 
+static int
+varnish_submit_gauge_with_target(const char *plugin_instance, /* {{{ */
+                                 const char *category, const char *target,
+                                 const char *type, const char *type_instance,
+                                 uint64_t gauge_value) {
+
+  return varnish_submit(plugin_instance, category, target, type, type_instance,
+                        (value_t){
+                            .gauge = (gauge_t)gauge_value,
+                        });
+} /* }}} int varnish_submit_gauge_with_target */
+
 static int varnish_submit_derive(const char *plugin_instance, /* {{{ */
                                  const char *category, const char *type,
                                  const char *type_instance,
                                  uint64_t derive_value) {
-  return varnish_submit(plugin_instance, category, type, type_instance,
+  return varnish_submit(plugin_instance, category, NULL, type, type_instance,
                         (value_t){
                             .derive = (derive_t)derive_value,
                         });
 } /* }}} int varnish_submit_derive */
 
+static int
+varnish_submit_derive_with_target(const char *plugin_instance, /* {{{ */
+                                  const char *category, const char *target,
+                                  const char *type, const char *type_instance,
+                                  uint64_t derive_value) {
+  return varnish_submit(plugin_instance, category, target, type, type_instance,
+                        (value_t){
+                            .derive = (derive_t)derive_value,
+                        });
+} /* }}} int varnish_submit_derive_with_target */
+
 #if HAVE_VARNISH_V3 || HAVE_VARNISH_V4 || HAVE_VARNISH_V5 || HAVE_VARNISH_V6
 static int varnish_monitor(void *priv,
                            const struct VSC_point *const pt) /* {{{ */
 {
   uint64_t val;
   const user_config_t *conf;
   const char *name;
+  //  const char *stat_category;
+  const char *stat_target;
 
   if (pt == NULL)
     return 0;
 
   conf = priv;
 
-#if HAVE_VARNISH_V5 || HAVE_VARNISH_V6
+#if HAVE_VARNISH_V6
+  /*
+   stats examples:
+   MAIN.threads
+   SMA.s0.c_req
+   SMA.Transient.c_req
+   VBE.vclName.backendName.req // where vclName changes every time a new vcl is
+   loaded
+  */
+  const int MAX_TOKENS = 4; // we expect 4 maximum tokens
+  char *token_array[MAX_TOKENS];
+  int i = 0;
+  int token_count = -1;
+  char *buffer = strdup(pt->name);
+  char *running = buffer;
+  do {
+    token_array[i] = strsep(&running, ".");
+    if (token_array[i] == NULL) {
+      token_count = i - 1;
```

Comment:
This is very surprising and hard to read: if you have four tokens, the tokens will be in `token_array[0..3]` and `token count` will be 2 – neither the last element nor the actual count. Please either set the token count to the actual count, i.e. `i + 1`, or rename the variable to `last_index` and set it to `i`.

Reply:
when we get a null token, we're one index too far, hence the minus one. but you're right, this is an index and not a count :)